core/src/main/java/org/apache/calcite/plan/volcano/RelSet.java

```diff
@@ -317,8 +284,8 @@ RelSubset getOrCreateSubset(
       subset.setDelivered();
     }
 
-    if (needsConverter && !planner.topDownOpt) {
-      addConverters(subset, required, true);
+    if (needsConverter) {
+      addConverters(subset, required, !planner.topDownOpt);
```

Comment:
hmm, that means whenever a new traitset is added, it will always add the enforcer if needed.

Reply:
Yes, it is. The converters act as bridges between the working subset and other subset. Currently, the OptimizeGroup task will schedule tasks for all the logical Rels in the RelSet while it skips physical Rels from other subsets. That's because we are confident that if a physical rel can be reach by a converter, the converter is already in the current working subset.

I was thinking of checking the upper bound before adding the enforcer. But I cannot decide which place is suitable for adding the enforcer back when the upper bound is changed. So I leave it for further optimizations.